ssm/variational.py

```diff
@@ -283,75 +334,138 @@ def sample_discrete_states(self):
-        return [block_tridiagonal_mean(prms["J_diag"], prms["J_lower_diag"], prms["h"], lower=True).reshape(np.shape(prms["h"]))
+        full_expectations = self.continuous_expectations
+        return [exp[1] for exp in full_expectations]
+
+    @property
+    def continuous_expectations(self):
+        return [kalman_info_smoother(prms["J_ini"], prms["h_ini"], 0,
+                                     prms["J_dyn_11"], prms["J_dyn_21"], prms["J_dyn_22"],
+                                     prms["h_dyn_1"], prms["h_dyn_2"], 0,
+                                     prms["J_obs"], prms["h_obs"], 0)
                 for prms in self.params]
 
     @property
     def mean(self):
         return list(zip(self.mean_discrete_states, self.mean_continuous_states))
 
-    def entropy(self, sample=None):
+    def _discrete_entropy(self):
+        negentropy = 0
+        for prms in self.params:
+            log_pi0 = np.log(prms["pi0"] + 1e-16) - logsumexp(prms["pi0"])
+            log_Ps = np.log(prms["Ps"] + 1e-16) - logsumexp(prms["Ps"],
+                                                            axis=1,
+                                                            keepdims=True)
+            (Ez, Ezzp1, normalizer) = hmm_expected_states(prms["pi0"],
+                                                          prms["Ps"],
+                                                          prms["log_likes"])
+
+            negentropy -= normalizer  # -log Z
+            negentropy += np.sum(Ez[0] * log_pi0)  # initial factor
+            negentropy += np.sum(Ez * prms["log_likes"])  # unitary factors
+            negentropy += np.sum(Ezzp1 * log_Ps)  # pairwise factors
+        return -negentropy
+
+    def _continuous_entropy(self, expectations):
+        negentropy = 0
+        Exs, ExxTs, ExxnTs, log_Zs = expectations
+        for prms, Ex, ExxT, ExxnT, log_Z in zip(self.params,
+                                                Exs,
+                                                ExxTs,
+                                                ExxnTs,
+                                                log_Zs):
+            # Pairwise terms
+            negentropy += np.sum(-0.5 * trace_product(prms["J_ini"], ExxT[0]))
+            negentropy += np.sum(-0.5 * trace_product(prms["J_dyn_11"], ExxT[:-1]))
+            negentropy += np.sum(-0.5 * trace_product(prms["J_dyn_22"], ExxT[1:]))
+            negentropy += np.sum(-0.5 * trace_product(prms["J_obs"], ExxT))
+            negentropy += np.sum(-1.0 * trace_product(prms["J_dyn_21"], ExxnT))
+
+            # Unary terms
+            negentropy += np.sum(prms["h_ini"] * Ex[0])
+            negentropy += np.sum(prms["h_dyn_1"] * Ex[:-1])
+            negentropy += np.sum(prms["h_dyn_2"] * Ex[1:])
+            negentropy += np.sum(prms["h_obs"] * Ex)
+
+            # Log normalizer
+            negentropy -= log_Z
+        return -negentropy
+
+    def _approx_continuous_entropy(self, sample):
```

Comment:
I don't see any need for this anymore.

Reply:
okay cool. I was keeping it because I thought it might be relevant for the "sticking the landing thing". If someone sets parameters_update="sgd" we would be differentiating through this entropy calculation. But I also would be in favor of taking it out for now for the sake of simplicity and readability -- we can always put it back later.

Reply:
Ah, right. That could be useful for reducing variance of the gradient calculations for the SGD 
M step.  I'm concerned that this function has way too many branching paths though.  Let's get rid of as many if/else blocks as we can.  

Concretely, I would suggest getting rid of the approximate continuous entropy for now and the Laplace EM + SGD parameter updates since we're not using it in any examples.  We can create an issue to revisit this combo at a later date, if desired.  (Just link to this point in the code so we don't have to write it from scratch if we do return to it.)

Reply:
While we're at it, I think we should collapse `stoch_mstep` and `exact_mstep` since I can't imagine ever doing a stochastic update when we could do an exact one instead.

Reply:
cool. I'm getting rid of these branching paths. Issue created here: https://github.com/slinderman/ssm/issues/98